adafruit_airlift/esp32.py

```diff
@@ -15,111 +14,195 @@
-        self._busy_and_cts = digitalio.DigitalInOut(esp_busy)
-        self._chip_select = digitalio.DigitalInOut(esp_cs)
-        self._uart = busio.UART(
-            esp_tx, esp_rx, baudrate=115200, timeout=0, receiver_buffer_size=512
-        )
+        # Used for Bluetooth mode.
+        self._uart = None
+        self._bleio_adapter = None
+        # Used for WiFi mode.
+        self._spi = None
+
+    def reset(self, mode, debug=False):
+        """Do hard reset of the ESP32.
+
+        :param mode: One of `ESP32.NOT_IN_USE`, `ESP32.BOOTLOADER`, `ESP32.BLUETOOTH`, `ESP32.WIFI`.
+        """
+        if mode not in ESP32._MODES:
+            raise ValueError("Invalid mode")
+
+        # GPIO0 high means boot from SPI flash.
+        # Low means go into bootloader mode.
+        self._gpio0_rts.switch_to_output(mode != ESP32.BOOTLOADER)
+
+        if mode == ESP32.NOT_IN_USE:
+            # Turn of ESP32 by holding reset line.
+            self._reset.switch_to_output(self._reset_high)
+            self._mode = mode
+            return
+
+        if mode == ESP32.BLUETOOTH:
+            self._chip_select.switch_to_output(False)
+        elif mode == ESP32.WIFI:
+            self._chip_select.switch_to_output(True)
+
+        # Initial mode. Changed if reset is successful.
+        self._mode = ESP32.NOT_IN_USE
 
-    def _reset_esp32(self):
         # Reset by toggling reset pin for 100ms
-        self._reset.value = self._reset_high
+        self._reset.switch_to_output(self._reset_high)
         time.sleep(0.1)
         self._reset.value = not self._reset_high
 
         #  Wait 1 second for startup.
         time.sleep(1.0)
 
+        if mode == ESP32.BOOTLOADER:
+            # No startup message expected.
+            return
+
         startup_message = b""
         while self._uart.in_waiting:  # pylint: disable=no-member
             more = self._uart.read()
             if more:
                 startup_message += more
 
-        return startup_message
+        if not startup_message:
+            raise RuntimeError("ESP32 did not respond with a startup message")
+        if debug:
+            try:
+                print(startup_message.decode("utf-8"))
+            except UnicodeError:
+                raise RuntimeError("Garbled ESP32 startup message") from UnicodeError
+
+        # Everything's fine. Remember mode.
+        self._mode = mode
 
-    def start_bluetooth(self, debug=False):
+    def start_bluetooth(self, uart=None, debug=False):
         """Set up the ESP32 in HCI Bluetooth mode, if it is not already doing something else.
-        Return a _bleio.Adapter.
+
+        :param uart busio.UART: Used for communication with the ESP32.
+          ``uart`` must have its ``baudrate = 115200``, ``timeout = 0``,
+          and ``receiver_buffer_size`` at least of size 512.
+          If not supplied, ``board.TX`` and ``board.RX`` are used to create a suitable UART.
```

Comment:
Why have this here instead of the constructor?

Reply:
Because the bluetooth and wifi functions use UART and SPI, respectively, and if you're only doing one of those functions, you don't need the busio device for the other function. E.g., if you're only doing wifi, you would never need to create a UART, and you might not want to, in case, for instance, you are short of SERCOM's.

Reply:
I think it's better to do the construct the UART yourself in `start` with pins given in the constructor. UART can't be shared so you can do it for bluetooth at least.

Are the SPI lines ever shared? If not, you can hide the SPI init as well. Otherwise I think it's ok to require it passed in.

Reply:
The Metro M4 Airlift and PyPortal uses the standard SPI lines (not dedicated lines) for SPI (There is some trickery with MISO). And if you were using an offboard ESP32 it could also be non-dedicated

I agree that it would be fine to just pass the UART TX and RX pins in, and construct that, since it would be dedicated. I'll change that.

Reply:
Why are you passing the pins and SPI into the `start_*` functions? I think they should be passed into the constructor along with the control pins because they are fixed when the device is created. It isn't dynamic as the code.py runs.

Reply:
I had factored out the mode-specific pins/SPI into `start_*`, but yes, I can do that. I originally did it in `start_*` so it would be more obvious if you gave them incorrectly.

Reply:
OK, passed pins and SPI to `__init__`, with `None` as default. UART is not created until needed for Bluetooth. `board.SPI()` is not referenced until/if needed for WiFi.

Moved `reset_high` arg next to `reset` pin arg, since it's associated, instead of leaving it dangling at the end of the arg list.

Added `debug=False` parameter to `start_wifi()` to match `start_bluetooth()` `debug=False`.